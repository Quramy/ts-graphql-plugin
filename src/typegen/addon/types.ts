import ts from 'typescript';
import type {
  GraphQLSchema,
  GraphQLObjectType,
  GraphQLUnionType,
  GraphQLInterfaceType,
  GraphQLScalarType,
} from 'graphql/type';
import type { DocumentNode, OperationDefinitionNode, FragmentDefinitionNode } from 'graphql/language';
import { OutputSource } from '../../ts-ast-util/types';

type GraphQLFragmentTypeConditionNamedType = GraphQLObjectType | GraphQLUnionType | GraphQLInterfaceType;

type ToStrict<T> = { [P in keyof T]-?: T[P] };
export type StrictAddon = ToStrict<TypeGenVisitorAddon>;

/**
 *
 * Creates Type Generator addon object.
 * This factory function is called back for each GraphQL template string.
 *
 */
export interface TypeGenAddonFactory {
  /**
   *
   * @param context - @see TypeGenVisitorAddonContext
   * @returns Addon
   *
   */
  (context: TypeGenVisitorAddonContext): TypeGenVisitorAddon;
}

export interface TypeGenVisitorAddonContext {
  /**
   *
   * Referenced GraphQL schema object.
   *
   */
  readonly schema: GraphQLSchema;

  /**
   *
   * Utility object to mutate output source file. @see OutputSource
   *
   */
  readonly source: OutputSource;

  /**
   *
   * Where the GraphQL template string is located. @see ExtractedInfo
   *
   */
  readonly extractedInfo: ExtractedInfo;
}

/**
 *
 * Represents where the template string including the GraphQL operation or fragments are located.
 *
 */
export interface ExtractedInfo {
  /**
   *
   * File path of the .ts file which contains the GraphQL template string.
   *
   */
  readonly fileName: string;

  /**
   *
   * TypeScript source file object which contains the GraphQL template string.
   *
   */
  readonly tsSourceFile: ts.SourceFile;

  /**
   *
   * Template string AST node corresponding the GraphQL operation or fragments.
   *
   */
  readonly tsTemplateNode: ts.NoSubstitutionTemplateLiteral | ts.TemplateExpression;
}

/**
 *
 * The addon object's methods are called back during the core type generator visits each GraphQL node in the template string.
 * And one addon object is generated for one template string.
 *
 * You don't have to implement all methods.
 *
 */
export interface TypeGenVisitorAddon {
  /**
   *
   * Reacts when the type generator visits GraphQL Scalar field.
   * This function can return corresponding TypeScript type node.
   *
   * @remarks
   * This function can not how to serialize/deserialize the custom scalar but how to map scalar to TypeScript type only.
   * You should consistence between runtime serialize behavior and type mapping.
   *
   * @example
   *
   * If your schema has a custom scalar type `Date` and your GraphQL client deserializes it to JavaScript native string,
   * you can implement this type mapping as the following:
   *
   * ```graphql
   * scalar Date
   * type Post {
   *   id: ID!
   *   publishedAt: Date!
   * }
   * ```
   *
   * ```ts
   * const factory: TypeGenAddonFactory = () => ({
   *   customScalar({ scalarType }) {
   *     if (scalarType.name === 'Date') {
   *       return ts.createKeywordTypeNode(ts.SyntaxKind.StringKeyword);
   *     }
   *   },
   * });
   * ```
   *
   * @param input - contains GraphQL Scalar type.
   * @return Corresponding TypeScript type node. If return undefined, the TypeScript type is determined by the core type generator.
   *
   */
  customScalar?: (input: CustomScalarInput) => CustomScalarOutput;

  /**
   *
   * Reacts when the type generator visits whole GraphQL document node.
   *
   * @param input - contains the GraphQL document AST node.
   *
   */
  document?: (input: DocumentInput) => void;

  /**
   *
   * Reacts when the type generator visits GraphQL operation definition node.
   * It's useful if you want to process query or variables types.
   *
   * @example
   *
   * ```ts
   * const factory: TypeGenAddonFactory = ({ source }) => ({
   *   operationDefinition({ tsResultNode }) {
   *     const additionalTypeStatement = createAwesomeType(tsResultNode);
   *     source.pushStatement(additionalTypeStatement);
   *   },
   * });
   * ```
   *
   * @param input - contains query/variable TypeScript type declaration node. @see OperationDefinionInput
   *
   */
  operationDefinition?: (input: OperationDefinionInput) => void;

  /**
   *
   * Reacts when the type generator visits GraphQL fragment definition node.
   *
   * @param input - contains fragment TypeScript type declaration node. @see FragmentDefinitionInput
   *
   */
  fragmentDefinition?: (input: FragmentDefinitionInput) => void;
}

/**
 *
 * Input object for `TypeGenVisitorAddon.customScalar` callback.
 *
 */
export interface CustomScalarInput {
  /**
   *
   * Definition of type scalar type.
   *
   */
  readonly scalarType: GraphQLScalarType;
}

/**
 *
 * Output object for `TypeGenVisitorAddon.customScalar` callback.
 *
 */
export type CustomScalarOutput = void | ts.TypeNode;

/**
 *
 * Input object for `TypeGenVisitorAddon.document` callback.
 *
 */
export interface DocumentInput {
  /**
   *
   * GraphQL document AST node.
   *
   */
  readonly graphqlNode: DocumentNode;
}

/**
 *
 * Input object for `TypeGenVisitorAddon.operationDefinition` callback.
 *
 */
export interface OperationDefinionInput {
  /**
   *
   * GraphQL AST node of this operation.
   *
   */
  readonly graqhqlNode: OperationDefinitionNode;

  /**
   *
   * Definition of the operation. One of Query or Mutation or Subscription
   *
   */
  readonly operationType: GraphQLObjectType;

  /**
   *
   * TypeScript AST node which represents the result type of this operation.
   *
   */
  readonly tsResultNode: ts.TypeAliasDeclaration;

  /**
   *
   * TypeScript AST node which represents the variable type of this operation.
   *
   */
  readonly tsVariableNode: ts.TypeAliasDeclaration;
}

/**
 *
 * Input object for `TypeGenVisitorAddon.fragmentDefinition` callback.
 *
 */
export interface FragmentDefinitionInput {
  /**
   *
   * GraphQL AST node of this fragment
   *
   */
  readonly graphqlNode: FragmentDefinitionNode;

  /**
   *
   * Named type(object or interface or union type) this fragment is specified on.
   *
   */
  readonly conditionType: GraphQLFragmentTypeConditionNamedType;

  /**
   *
   * TypeScript AST node which represents this fragment type.
   *
   */
  readonly tsNode: ts.TypeAliasDeclaration;
}
