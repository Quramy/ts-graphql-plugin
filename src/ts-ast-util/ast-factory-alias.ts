import ts from 'typescript';

export const astf = ts.factory;
