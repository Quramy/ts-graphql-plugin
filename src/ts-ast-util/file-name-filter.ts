import _path from 'node:path';

import { globToRegExp } from '../string-util/glob-to-regexp';

export function createFileNameFilter({
  specs,
  projectName,
  _forceWin32 = false,
}: {
  specs: string[] | undefined;
  projectName: string;
  _forceWin32?: boolean;
}) {
  const path = _forceWin32 ? _path.win32 : _path;

  const testers = (specs ?? []).map(pattern => {
    if (pattern.includes('*') || pattern.includes('?')) {
      const regexp = globToRegExp(pattern);
      return (normalized: string) => regexp.test(normalized);
    } else {
      const dirOrFileName = pattern[pattern.length - 1] === '/' ? pattern.slice(0, pattern.length - 1) : pattern;
      return (normalized: string) => normalized === dirOrFileName || normalized.startsWith(dirOrFileName + '/');
    }
  });

  const projectRootDirName = path.dirname(projectName);

  const match = (fileName: string) => {
    const normalized = path.relative(projectRootDirName, fileName).replace(/\\/g, '/');
    return testers.some(tester => tester(normalized));
  };

  return match;
}
