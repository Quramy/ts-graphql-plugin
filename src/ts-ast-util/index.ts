import ts from 'typescript/lib/tsserverlibrary';

export * from './types';

export * from './template-expression-resolver';

export { createScriptSourceHelper } from './script-source-helper';

export type TagCondition = string;

export function findNode(sourceFile: ts.SourceFile, position: number): ts.Node | undefined {
  function find(node: ts.Node): ts.Node | undefined {
    if (position >= node.getStart() && position < node.getEnd()) {
      return ts.forEachChild(node, find) || node;
    }
  }
  return find(sourceFile);
}

export function findAllNodes(sourceFile: ts.SourceFile, cond: (n: ts.Node) => boolean): ts.Node[] {
  const result: ts.Node[] = [];
  function find(node: ts.Node) {
    if (cond(node)) {
      result.push(node);
      return;
    } else {
      ts.forEachChild(node, find);
    }
  }
  find(sourceFile);
  return result;
}

export function isTagged(node: ts.Node, condition: TagCondition) {
  if (!node || !node.parent) return false;
  if (!ts.isTaggedTemplateExpression(node.parent)) return false;
  const tagNode = node.parent;
  return tagNode.tag.getText() === condition;
}

export function isTemplateLiteralTypeNode(node: ts.Node): node is ts.TemplateLiteralTypeNode {
  // ts.isNoSubstitutionTemplateLiteral exists TypeScript >= 4.1
  return typeof ts.isTemplateLiteralTypeNode === 'function' && ts.isTemplateLiteralTypeNode(node);
}
