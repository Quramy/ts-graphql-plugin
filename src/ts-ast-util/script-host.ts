import ts from 'typescript';

export class ScriptHost implements ts.LanguageServiceHost {
  private readonly _fileMap = new Map<string, string | undefined>();
  private readonly _fileVersionMap = new Map<string, number>();

  constructor(private readonly _currentDirectory: string, private readonly _compilerOptions: ts.CompilerOptions) {}

  readFile(fileName: string) {
    const hit = this._fileMap.get(fileName);
    if (hit != null) return hit;
    return this.loadFromFileSystem(fileName);
  }

  loadFromFileSystem(fileName: string) {
    const content = ts.sys.readFile(fileName, 'uts8');
    this._updateFile(fileName, content);
    return content;
  }

  getCurrentDirectory() {
    return this._currentDirectory;
  }

  getScriptSnapshot(fileName: string) {
    const file = this._fileMap.get(fileName);
    if (file == null) return;
    return ts.ScriptSnapshot.fromString(file);
  }

  getScriptVersion(fileName: string) {
    const version = this._fileVersionMap.get(fileName);
    if (!version) return '0';
    return version + '';
  }

  getScriptFileNames() {
    return [...this._fileMap.keys()];
  }

  getCompilationSettings() {
    return this._compilerOptions;
  }

  getDefaultLibFileName(opt: ts.CompilerOptions) {
    return ts.getDefaultLibFileName(opt);
  }

  // Sinse ts 4.7, imeplement required.
  // Delegating to `ts.sys.fileExists(path)` is not for me, but I don't know whether the following implementation is correct.
  fileExists() {
    return false;
  }

  protected _updateFile(fileName: string, content: string | undefined) {
    this._fileMap.set(fileName, content);
    const currentVersion = this._fileVersionMap.get(fileName) || 0;
    this._fileVersionMap.set(fileName, currentVersion + 1);
  }
}
