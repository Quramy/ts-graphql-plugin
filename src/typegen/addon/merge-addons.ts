import {
  TypeGenVisitorAddon,
  StrictAddon,
  CustomScalarInput,
  CustomScalarOutput,
  DocumentInput,
  OperationDefinionInput,
  FragmentDefinitionInput,
} from './types';

export function mergeAddons(addonList: (TypeGenVisitorAddon | undefined)[]) {
  const addon: StrictAddon = {
    customScalar: (input: CustomScalarInput) => {
      return addonList.reduce((acc: CustomScalarOutput | undefined, addon) => {
        return addon?.customScalar ? acc || addon?.customScalar(input) : acc;
      }, undefined);
    },

    document(input: DocumentInput) {
      return addonList.forEach(addon => addon?.document?.(input));
    },

    operationDefinition(input: OperationDefinionInput) {
      return addonList.forEach(addon => addon?.operationDefinition?.(input));
    },

    fragmentDefinition(input: FragmentDefinitionInput) {
      return addonList.forEach(addon => addon?.fragmentDefinition?.(input));
    },
  };

  return addon;
}
