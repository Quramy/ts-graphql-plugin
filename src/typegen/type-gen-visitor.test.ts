import { TypeGenVisitor, TypeGenError } from './type-gen-visitor';
import ts from 'typescript';
import { parse, buildSchema } from 'graphql';
import { mergeAddons } from './addon/merge-addons';
import { createOutputSource } from '../ts-ast-util';

function generateAstAndPrint({ schemaSDL, documentContent }: { schemaSDL: string; documentContent: string }) {
  const schema = buildSchema(schemaSDL);
  const documentNode = parse(documentContent);
  const source = new TypeGenVisitor({ schema }).visit(documentNode, {
    outputSource: createOutputSource({ outputFileName: 'out.ts' }),
    addon: mergeAddons([]),
  });
  const printer = ts.createPrinter({ newLine: ts.NewLineKind.LineFeed, removeComments: true });
  return printer.printFile(source);
}

describe('typegen', () => {
  describe('variable type generation', () => {
    it('should throw error when reference to not existing input type', () => {
      expect(() =>
        generateAstAndPrint({
          schemaSDL: `
              type Query {
                hello(var: String!): String!
              }
            `,
          documentContent: `
              query MyQuery (
                $idVar: UnknowType,
              ) {
                hello(var: $var)
              }
            `,
        }),
      ).toThrowError(TypeGenError);
    });

    it('should gen type from scalar types', () => {
      const result = generateAstAndPrint({
        schemaSDL: `
          scalar MyScalar
          type Query {
            idInput(idValue: ID!): String!
            strInput(strValue: String!): String!
            intInput(intValue: Int!): String!
            floatInput(floatValue: Float!): String!
            boolInput(boolValue: Boolean!): String!
            customScalarInput(value: MyScalar!): String!
          }
        `,
        documentContent: `
          query MyQuery (
            $idVar: ID!,
            $strVar: String!,
            $intVar: Int!,
            $floatVar: Float!,
            $boolVar: Boolean!,
            $scalarVar: MyScalar!,
          ) {
            idInput(idValue: $idVar)
            strInput(strValue: $strVar)
            intInput(intValue: $intVar)
            floatInput(floatValue: $floatVar)
            boolInput(boolValue: $boolVar)
            customScalarInput(value: $scalarVar)
          }
        `,
      });
      expect(result).toMatchSnapshot();
    });

    it('should gen type from enum', () => {
      const result = generateAstAndPrint({
        schemaSDL: `
          enum Color {
            RED
            BLUE
            GREEN
          }
          type Query {
            enumInput(color: Color): String!
          }
        `,
        documentContent: `
          query MyQuery ($color: Color!) {
            enumInput(color: $color)
          }
        `,
      });
      expect(result).toMatchSnapshot();
    });

    it('should gen type from input object type', () => {
      const result = generateAstAndPrint({
        schemaSDL: `
          input Vector {
            x: Float!
            y: Float!
          }
          input Line {
            start: Vector!
            end: Vector!
          }
          type Query {
            lineInput(line: Line!): Float!
          }
        `,
        documentContent: `
          query MyQuery ($line: Line!) {
            lineInput(line: $line)
          }
        `,
      });
      expect(result).toMatchSnapshot();
    });

    it('should gen type with correct list/nonNull modifiers', () => {
      const result = generateAstAndPrint({
        schemaSDL: `
          type Query {
            nullableField(value: String): String!
            strictField(value: String): String!
            nullableFieldNullableList(value: [String]): String!
            nullableFieldStrictList(value: [String]!): String!
            strictFieldNullableList(value: [String!]): String!
            strictFieldStrictList(value: [String!]!): String!
            nullableNestedList(value: [[String]]): String!
            strictNestedList(value: [[String!]!]!): String!
          }
        `,
        documentContent: `
          query MyQuery (
            $var1: String,
            $var2: String!,
            $var3: [String],
            $var4: [String]!,
            $var5: [String!],
            $var6: [String!]!,
            $var7: [[String]],
            $var8: [[String!]!]!,
          ){
            nullableField(value: $var1)
            strictField(value: $var2)
            nullableFieldNullableList(value: $var3)
            nullableFieldStrictList(value: $var4)
            strictFieldNullableList(value: $var5)
            strictFieldStrictList(value: $var6)
            nullableNestedList(value: $var7)
            strictNestedList(value: $var8)
          }
        `,
      });
      expect(result).toMatchSnapshot();
    });

    it('should gen optional type with default value variable', () => {
      const result = generateAstAndPrint({
        schemaSDL: `
          type Query {
            hello(count: Int!): String!
          }
        `,
        documentContent: `
          query MyQuery ($count: Int = 10) {
            hello(count: $count)
          }
        `,
      });
      expect(result).toMatchSnapshot();
    });

    it('should work even if the input types have circular references', () => {
      const result = generateAstAndPrint({
        schemaSDL: `
          input User {
            posts: [Post!]!
          }
          input Post {
            author: User!
          }
          type Mutation {
            addPost(post: Post!): Boolean
          }
        `,
        documentContent: `
          mutation AddPost($post: Post!) {
            addPost(post: $post)
          }
        `,
      });
      expect(result).toMatchSnapshot();
    });
  });

  describe('result type generation', () => {
    describe('output types pattern', () => {
      it('should throw error when reference to Query but schema does not have Query type', () => {
        expect(() =>
          generateAstAndPrint({
            schemaSDL: `
              type Mutation {
                hello: String!
              }
            `,
            documentContent: `
              query MyQuery {
                hello
              }
            `,
          }),
        ).toThrowError(TypeGenError);
      });

      it('should throw error when reference to Mutation but schema does not have Mutation type', () => {
        expect(() =>
          generateAstAndPrint({
            schemaSDL: `
              type Query {
                hello: String!
              }
            `,
            documentContent: `
              mutation MyMutaion {
                hello
              }
            `,
          }),
        ).toThrowError(TypeGenError);
      });

      it('should throw error when reference to Subscription but schema does not have Subscription type', () => {
        expect(() =>
          generateAstAndPrint({
            schemaSDL: `
              type Query {
                hello: String!
              }
            `,
            documentContent: `
              subscription MySubscription {
                hello
              }
            `,
          }),
        ).toThrowError(TypeGenError);
      });

      it('should throw error when reference to not existing field', () => {
        expect(() =>
          generateAstAndPrint({
            schemaSDL: `
              type Query {
                hello: String!
              }
            `,
            documentContent: `
              query MyQuery {
                goodBye
              }
            `,
          }),
        ).toThrowError(TypeGenError);
      });

      it('should gen type from built-in scalar types', () => {
        const result = generateAstAndPrint({
          schemaSDL: `
            type Query {
              idField: ID!
              stringField: String!
              intField: Int!
              floatField: Float!
              boolField: Boolean!
            }
          `,
          documentContent: `
            query MyQuery {
              idField
              stringField
              intField
              floatField
              boolField
            }
          `,
        });
        expect(result).toMatchSnapshot();
      });

      it('should gen type from enum', () => {
        const result = generateAstAndPrint({
          schemaSDL: `
            enum Color {
              RED
              BLUE
              GREEN
            }
            type Query {
              color: Color
            }
          `,
          documentContent: `
            query MyQuery {
              color
            }
          `,
        });
        expect(result).toMatchSnapshot();
      });

      it('should gen type from object type', () => {
        const result = generateAstAndPrint({
          schemaSDL: `
            type Query {
              hello: String!
            }
          `,
          documentContent: `
            query MyQuery {
              hello
            }
          `,
        });
        expect(result).toMatchSnapshot();
      });

      it('should gen type from interface type', () => {
        const result = generateAstAndPrint({
          schemaSDL: `
            interface Node {
              id: ID!
            }

            type Query {
              node: Node!
            }
          `,
          documentContent: `
            query MyQuery {
              node {
                id
              }
            }
          `,
        });
        expect(result).toMatchSnapshot();
      });

      it('should throw error when direct selection on union type', () => {
        expect(() =>
          generateAstAndPrint({
            schemaSDL: `
            type User {
              id: ID!
              name: String!
            }
            type Page {
              id: ID!
              body: String!
            }
            union Item = User | Page
            type Query {
              item: Item!
            }
          `,
            documentContent: `
            query MyQuery {
              item {
                id
              }
            }
          `,
          }),
        ).toThrowError(TypeGenError);
      });

      it('should gen type with correct list/nonNull modifiers', () => {
        const result = generateAstAndPrint({
          schemaSDL: `
            type Query {
              nullableField: String
              strictField: String!
              nullableFieldNullableList: [String]
              nullableFieldStrictList: [String]!
              strictFieldNullableList: [String!]
              strictFieldStrictList: [String!]!
              nullableNestedList: [[String]]
              strictNestedList: [[String!]!]!
            }
          `,
          documentContent: `
            query MyQuery {
              nullableField
              strictField
              nullableFieldNullableList
              nullableFieldStrictList
              strictFieldNullableList
              strictFieldStrictList
              nullableNestedList
              strictNestedList
            }
          `,
        });
        expect(result).toMatchSnapshot();
      });

      it('should gen type with field alias', () => {
        const result = generateAstAndPrint({
          schemaSDL: `
            type Query {
              hello: String!
            }
          `,
          documentContent: `
            query MyQuery {
              greeting: hello
            }
          `,
        });
        expect(result).toMatchSnapshot();
      });
    });

    describe('definition node pattern', () => {
      it('should gen type from query operation def', () => {
        const result = generateAstAndPrint({
          schemaSDL: `
            type Query {
              hello: String!
            }
          `,
          documentContent: `
            query {
              hello
            }
          `,
        });
        expect(result).toMatchSnapshot();
      });

      it('should gen type from mutation operation def', () => {
        const result = generateAstAndPrint({
          schemaSDL: `
            type Mutation {
              like(count: Int!): String!
            }
          `,
          documentContent: `
            mutation MyMutaion {
              like(count: 1)
            }
          `,
        });
        expect(result).toMatchSnapshot();
      });

      it('should gen type from subscription operation def', () => {
        const result = generateAstAndPrint({
          schemaSDL: `
            type Subscription {
              issue: String!
            }
          `,
          documentContent: `
            subscription MySubscription {
              issue
            }
          `,
        });
        expect(result).toMatchSnapshot();
      });

      it('should gen type from fragment def', () => {
        const result = generateAstAndPrint({
          schemaSDL: `
            type Query {
              hello: String
            }
          `,
          documentContent: `
            fragment MyFragment on Query {
              hello
            }
          `,
        });
        expect(result).toMatchSnapshot();
      });
    });

    describe('fragment spread reference', () => {
      it('should gen type reference from fragment spread on operation', () => {
        const result = generateAstAndPrint({
          schemaSDL: `
            type Query {
              hello: String!
              bye: String!
            }
          `,
          documentContent: `
            fragment MyFragment on Query {
              hello
            }
            query MyQuery {
              bye
              ...MyFragment
            }
          `,
        });
        expect(result).toMatchSnapshot();
      });

      it('should gen type reference from fragment spread on fragment def', () => {
        const result = generateAstAndPrint({
          schemaSDL: `
            type User {
              name: String!
              age: Int!
            }
          `,
          documentContent: `
            fragment A on User {
              name
            }
            fragment B on User {
              ...A
              age
            }
          `,
        });
        expect(result).toMatchSnapshot();
      });

      it('should gen type reference from fragment spread on field selection', () => {
        const result = generateAstAndPrint({
          schemaSDL: `
            type User {
              id: ID!
              name: String!
              age: Int!
            }
            type Query {
              users: [User!]!
            }
          `,
          documentContent: `
            fragment MyFragment on User {
              name
              age
            }
            query MyQuery {
              users {
                id
                ...MyFragment
              }
            }
          `,
        });
        expect(result).toMatchSnapshot();
      });

      it('should gen type from inline fragment without type condition', () => {
        const result = generateAstAndPrint({
          schemaSDL: `
            type Query {
              id: ID!
              hello: String!
            }
          `,
          documentContent: `
            query MyQuery {
              id
              ... {
                hello
              }
            }
          `,
        });
        expect(result).toMatchSnapshot();
      });

      it('should gen type from inline fragment with type condition', () => {
        const result = generateAstAndPrint({
          schemaSDL: `
            type Query {
              id: ID!
              hello: String!
            }
          `,
          documentContent: `
            query MyQuery {
              id
              ... on Query {
                hello
              }
            }
          `,
        });
        expect(result).toMatchSnapshot();
      });

      it('should gen union of type references from fragment concrete type condition', () => {
        const result = generateAstAndPrint({
          schemaSDL: `
            interface Node {
              id: ID!
            }
            type User implements Node {
              id: ID!
              name: String!
            }
            type Page implements Node {
              id: ID!
              body: String!
            }
            union Item = User | Page
            type Query {
              item: Item!
            }
          `,
          documentContent: `
            fragment NodeFragment on Node {
              id
            }
            fragment UserFragment on User {
              name
            }
            fragment PageFragment on Page {
              body
            }
            query MyQuery {
              item {
                ...UserFragment
                ...NodeFragment
                ...PageFragment
              }
            }
          `,
        });
        expect(result).toMatchSnapshot();
      });

      it('should gen union object literal type from inline fragment concrete type condition', () => {
        const result = generateAstAndPrint({
          schemaSDL: `
            interface Node {
              id: ID!
            }
            type User implements Node {
              id: ID!
              name: String!
            }
            type Page implements Node {
              id: ID!
              body: String!
            }
            union Item = User | Page
            type Query {
              item: Item!
            }
          `,
          documentContent: `
            query MyQuery {
              item {
                ... on Node {
                  id
                }
                ... on User {
                  name
                }
                ... on Page {
                  body
                }
              }
            }
          `,
        });
        expect(result).toMatchSnapshot();
      });
    });

    describe('__typename field', () => {
      it('should gen __typename type from object type', () => {
        const result = generateAstAndPrint({
          schemaSDL: `
            type Query {
              hello: String!
            }
          `,
          documentContent: `
            query MyQuery {
              __typename
            }
          `,
        });
        expect(result).toMatchSnapshot();
      });

      it('should gen __typename type from union type', () => {
        const result = generateAstAndPrint({
          schemaSDL: `
            type User {
              id: ID!
              name: String!
            }
            type Page {
              id: ID!
              body: String!
            }
            union Item = User | Page
            type Query {
              item: Item!
            }
          `,
          documentContent: `
            query MyQuery {
              item {
                __typename
              }
            }
          `,
        });
        expect(result).toMatchSnapshot();
      });

      it('should gen __typename type from interface type', () => {
        const result = generateAstAndPrint({
          schemaSDL: `
            interface Node {
              id: ID!
            }

            type Query {
              node: Node!
            }
          `,
          documentContent: `
            query MyQuery {
              node {
                __typename
              }
            }
          `,
        });
        expect(result).toMatchSnapshot();
      });
    });
  });
});
