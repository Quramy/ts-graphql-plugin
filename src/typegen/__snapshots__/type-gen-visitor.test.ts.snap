// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`typegen result type generation __typename field should gen __typename type from interface type 1`] = `
"export type MyQuery = {
    node: {
        __typename: string;
    };
};
export type MyQueryVariables = {};
"
`;

exports[`typegen result type generation __typename field should gen __typename type from object type 1`] = `
"export type MyQuery = {
    __typename: \\"Query\\";
};
export type MyQueryVariables = {};
"
`;

exports[`typegen result type generation __typename field should gen __typename type from union type 1`] = `
"export type MyQuery = {
    item: {
        __typename: \\"User\\" | \\"Page\\";
    };
};
export type MyQueryVariables = {};
"
`;

exports[`typegen result type generation definition node pattern should gen type from fragment def 1`] = `
"export type MyFragment = {
    hello: string | null;
};
"
`;

exports[`typegen result type generation definition node pattern should gen type from mutation operation def 1`] = `
"export type MyMutaion = {
    like: string;
};
export type MyMutaionVariables = {};
"
`;

exports[`typegen result type generation definition node pattern should gen type from query operation def 1`] = `
"export type QueryResult = {
    hello: string;
};
export type QueryVariables = {};
"
`;

exports[`typegen result type generation definition node pattern should gen type from subscription operation def 1`] = `
"export type MySubscription = {
    issue: string;
};
export type MySubscriptionVariables = {};
"
`;

exports[`typegen result type generation fragment spread reference should gen type from inline fragment with type condition 1`] = `
"export type MyQuery = {
    id: string;
} & ({
    hello: string;
});
export type MyQueryVariables = {};
"
`;

exports[`typegen result type generation fragment spread reference should gen type from inline fragment without type condition 1`] = `
"export type MyQuery = {
    id: string;
    hello: string;
};
export type MyQueryVariables = {};
"
`;

exports[`typegen result type generation fragment spread reference should gen type reference from fragment spread on field selection 1`] = `
"export type MyFragment = {
    name: string;
    age: number;
};
export type MyQuery = {
    users: ({
        id: string;
    } & MyFragment)[];
};
export type MyQueryVariables = {};
"
`;

exports[`typegen result type generation fragment spread reference should gen type reference from fragment spread on fragment def 1`] = `
"export type A = {
    name: string;
};
export type B = {
    age: number;
} & A;
"
`;

exports[`typegen result type generation fragment spread reference should gen type reference from fragment spread on operation 1`] = `
"export type MyFragment = {
    hello: string;
};
export type MyQuery = {
    bye: string;
} & MyFragment;
export type MyQueryVariables = {};
"
`;

exports[`typegen result type generation fragment spread reference should gen union object literal type from inline fragment concrete type condition 1`] = `
"export type MyQuery = {
    item: ({
        id: string;
    }) & (({
        name: string;
    }) | ({
        body: string;
    }));
};
export type MyQueryVariables = {};
"
`;

exports[`typegen result type generation fragment spread reference should gen union of type references from fragment concrete type condition 1`] = `
"export type NodeFragment = {
    id: string;
};
export type UserFragment = {
    name: string;
};
export type PageFragment = {
    body: string;
};
export type MyQuery = {
    item: NodeFragment & (UserFragment | PageFragment);
};
export type MyQueryVariables = {};
"
`;

exports[`typegen result type generation output types pattern should gen type from built-in scalar types 1`] = `
"export type MyQuery = {
    idField: string;
    stringField: string;
    intField: number;
    floatField: number;
    boolField: boolean;
};
export type MyQueryVariables = {};
"
`;

exports[`typegen result type generation output types pattern should gen type from enum 1`] = `
"export type MyQuery = {
    color: (\\"RED\\" | \\"BLUE\\" | \\"GREEN\\") | null;
};
export type MyQueryVariables = {};
"
`;

exports[`typegen result type generation output types pattern should gen type from interface type 1`] = `
"export type MyQuery = {
    node: {
        id: string;
    };
};
export type MyQueryVariables = {};
"
`;

exports[`typegen result type generation output types pattern should gen type from object type 1`] = `
"export type MyQuery = {
    hello: string;
};
export type MyQueryVariables = {};
"
`;

exports[`typegen result type generation output types pattern should gen type with correct list/nonNull modifiers 1`] = `
"export type MyQuery = {
    nullableField: string | null;
    strictField: string;
    nullableFieldNullableList: (string | null)[] | null;
    nullableFieldStrictList: (string | null)[];
    strictFieldNullableList: string[] | null;
    strictFieldStrictList: string[];
    nullableNestedList: ((string | null)[] | null)[] | null;
    strictNestedList: string[][];
};
export type MyQueryVariables = {};
"
`;

exports[`typegen result type generation output types pattern should gen type with field alias 1`] = `
"export type MyQuery = {
    greeting: string;
};
export type MyQueryVariables = {};
"
`;

exports[`typegen variable type generation should gen optional type with default value variable 1`] = `
"export type MyQuery = {
    hello: string;
};
export type MyQueryVariables = {
    count?: number | null;
};
"
`;

exports[`typegen variable type generation should gen type from enum 1`] = `
"export type MyQuery = {
    enumInput: string;
};
export type MyQueryVariables = {
    color: \\"RED\\" | \\"BLUE\\" | \\"GREEN\\";
};
"
`;

exports[`typegen variable type generation should gen type from input object type 1`] = `
"type VectorInputType = {
    x: number;
    y: number;
};
type LineInputType = {
    start: VectorInputType;
    end: VectorInputType;
};
export type MyQuery = {
    lineInput: number;
};
export type MyQueryVariables = {
    line: LineInputType;
};
"
`;

exports[`typegen variable type generation should gen type from scalar types 1`] = `
"export type MyQuery = {
    idInput: string;
    strInput: string;
    intInput: string;
    floatInput: string;
    boolInput: string;
    customScalarInput: string;
};
export type MyQueryVariables = {
    idVar: string;
    strVar: string;
    intVar: number;
    floatVar: number;
    boolVar: boolean;
    scalarVar: any;
};
"
`;

exports[`typegen variable type generation should gen type with correct list/nonNull modifiers 1`] = `
"export type MyQuery = {
    nullableField: string;
    strictField: string;
    nullableFieldNullableList: string;
    nullableFieldStrictList: string;
    strictFieldNullableList: string;
    strictFieldStrictList: string;
    nullableNestedList: string;
    strictNestedList: string;
};
export type MyQueryVariables = {
    var1?: string | null;
    var2: string;
    var3?: (string | null)[] | null;
    var4: (string | null)[];
    var5?: string[] | null;
    var6: string[];
    var7?: ((string | null)[] | null)[] | null;
    var8: string[][];
};
"
`;

exports[`typegen variable type generation should work even if the input types have circular references 1`] = `
"type UserInputType = {
    posts: PostInputType[];
};
type PostInputType = {
    author: UserInputType;
};
export type AddPost = {
    addPost: boolean | null;
};
export type AddPostVariables = {
    post: PostInputType;
};
"
`;
